### PR TITLE
Wr/convert scripts

### PR DESCRIPTION
**Pull request by WillieRuemmele**
### What does this PR do?
adds a `convert script` command to help people migrate to `sf`

### What issues does this PR fix or reference?
@W-12942991@
https://github.com/forcedotcom/cli/issues/2071

--- 
QA Setup:

`sfdx/sf plugins link .` this repo
`sfdx dev convert script`

running with `./bin/dev` won't result in a complete command list and will fail to convert most commands

### CODE REVIEW

File: messages/convert.script.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+Description of a flag.
+
+# flags.no-prompt.summary
+
+Don't prompt for replacements
+
+# examples
+
+- <%= config.bin %> <%= command.id %> convert script --script myScript.yml
+
```
Comment (iowillhoit) [inline messages/convert.script.md:20]: Replace the boilerplate values in this file ^

File: src/commands/convert/script.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      try {
+        // if the line looks like it's a valid sfdx command
+        if (line.match(/sfdx \w+/g)?.length && line.includes(':') && !line.startsWith('#')) {
+          const commandId = line.split('sfdx ')[1]?.split(' ')[0];
+
+          const replacement = this.findReplacement(manifest, commandId);
+
+          // meaning it's deprecated to multiple commands and can't be replaced inline
+          if (Manifest[commandId]?.deprecationOptions?.to.includes('/')) {
+            this.warn(`Cannot determine appropriate replacement for ${commandId}`);
```
Comment (iowillhoit) [inline src/commands/convert/script.ts:75]: Might me nice to offer the line number here also

File: src/commands/convert/script.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        }
+      } catch (e) {
+        line = line.replace('sfdx ', 'sf ').replace(' -u ', ' --target-org ').replace(' -v ', ' --target-dev-hub');
+        line = line.concat(' # ERROR converting this line, human intervention required');
+        data.push(line);
+      }
+    }
+    // write the new script as a new file with a `-converted` suffix
+    fs.writeFileSync(
+      path.basename(flags.script, path.extname(flags.script)) + '-converted' + path.extname(flags.script),
```
[the remark this answers is not in the document]
Comment (WillieRuemmele) [reply src/commands/convert/script.ts:123]: ok - I just changed that yesterday based on Juliet's feedback. I'll add a message

**Comment by iowillhoit** [comment]
A few thoughts and questions:
- Is there a reason this new command was not added to the existing `src/dev/convert` directory? 
- Should we move the new `Manifest.ts` file alongside the `convert/script.ts` file since that is all it is used for? 
- I am concerned the `Manifest.ts` is going to get out of date and provide incorrect suggestions. What do you think? Is it worth trying to regenerate the oclif manifest as a pre-push or in a cron in GHA?

File: src/commands/convert/script.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      }
+    }
+    // write the new script as a new file with a `-converted` suffix
+    const output = path.basename(flags.script, path.extname(flags.script)) + '-converted' + path.extname(flags.script);
+    this.log(messages.getMessage('success', [output]));
+    fs.writeFileSync(output, data.join(os.EOL));
+  }
+
+  private async smartConfirm(message: string, prompt = true): Promise<boolean> {
+    return prompt ? await this.confirm(message, 100000) : true;
```
Comment (iowillhoit) [inline src/commands/convert/script.ts:124]: This timeout might still be a little short. I was curious about a suggestion and went to look if it was correct and it had timed out by the time I got back. Maybe 3 minutes? 180000

**Comment by iowillhoit** [comment]
Minor QA
- Also log a warning to stdout when "human intervention required", otherwise it looks like it converts successfully.
  - Maybe also log the `deprecationOptions.message` if it exists?
- Wrap commands in quotes to make them easier to read (especially space delimited)
- Add "command" and "flag" to the prompts?
- Change 'replacing' to 'replace' since it's a question
- Before
 ```
replace force:org:create for force org create Yes
? 	replacing f for definitionfile Yes
? 	replacing a for setalias Yes
```
- After
```
replace command "force:org:create" for "force org create" Yes
? 	replace flag "f" for "definitionfile" Yes
? 	replace flag "a" for "setalias" Yes
```

**Comment by iowillhoit** [comment]
## QA NOTES

- 🟡 Any command with `-h` or `--help` will fail
- 🔴 Any command with `--json` will fail
- 🔴 Any command with `--verbose` will fail
- 🟡 For some reason this convert is failing (seems right to me)
  - `sfdx force:package:install --package 0Ho4p0000004D99CAE`
  - `sf package install --package 0Ho4p0000004D99CAE # ERROR converting this line, human intervention required`
- 🟢 Works on a non-installed JIT plugin
- 🟢 The `--no-prompt` flag works
- 🟡 If the command you are converting already has spaces, `sfdx` is not converted to `sf`
  - `sfdx scanner rule describe` remains `sfdx scanner rule describe`
- 🔴 Oh, it looks like all commands with "long flags" are failing
  - Another `sfdx force:org:list --all`

FWIW, I ran this on my `.zsh-history` to get a wide variety of commands. Might be useful

Here are some additions to put in the `script.sh` for testing:

```
sfdx force:source:status --json
sfdx force:org:list
sfdx force:org:list -h
sfdx force:org:list --json
sfdx force:org:list --verbose
sfdx force:org:list --all
sfdx scanner:rule:list
sfdx scanner rule describe
sfdx force:package:install --package 04t1I0000000X0P
```

File: messages/dev.convert.script.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
+# summary
+
+Convert a script file that contains deprecated sfdx-style commands to use the new sf-style commands instead. 
+
+# description
+
+Important: Use this command only to get started on the sfdx->sf script migration. We don't guarantee that the new sf-style command replacements work correctly or as you expect. You must test, and probably update, the new script before putting it into production. We also don't guarantee that the JSON results are the same as before. 
+
+This command can convert a large part of your script, but possibly not all. There are some sfdx-style commands that don't have an obvious sf-style equivalent. In this case, this command doesn't replace the sfdx-style command but instead adds a comment to remind you that you must convert it manually.
```
Comment (jshackell-sfdc) [inline messages/dev.convert.script.md:9]:
```suggestion
This command can convert a large part of your script, but possibly not all. There are some sfdx-style commands that don't have an obvious sf-style equivalent. In this case, this command doesn't replace the sfdx-style command but instead adds a comment to remind you that you must convert it manually. See the Salesforce CLI Command Reference for migration information about each deprecated sfdx-style command: https://developer.salesforce.com/docs/atlas.en-us.sfdx_cli_reference.meta/sfdx_cli_reference/cli_reference.htm.
```

**Review by iowillhoit — APPROVED** [review]

**Review by peternhale — CHANGES_REQUESTED** [review]

**Pull request merged**